### ISSUE TITLE
Generic Types for Abstraction

### ISSUE THREAD

**Issue by MatheusFerreira71** [prisma/prisma#24789]
## Problem

I'd like to develop a Abstraction for my API's repository, using prisma and typescript, but when I'm typifying my variables I can't go further because prisma generates a diferente type for evey model in the schema and those types doesn't have a generic type (e.g Model<User>) or it doesn't extend any default type (e.g UserDelegate<ExtArgs extends $Extensions.InternalArgs = $Extensions.DefaultArgs> extends DefaultDelegate)

## Suggested solution

Create generic types for operation types (e.g export type UserDelegate = Delegate<User>)

## Alternatives

Create default types for extension (e.g UserDelegate extends DefaultDelegate)

## Additional context

Below is a example of what I'm building, if there is already a way to do this, please forgive me and post the link of the anwser somewhere. If needed to update the source code I'm willing to do it, I'll just need some guidance.

----------------------------------------------------------------------------------------------------------------------------------------------

import prisma from "../../prisma/prisma";

interface Pagination<T> {
  content: T[];
  count: number;
}

export default class Repository<T, R> {
  constructor(private repository: R) {
    this.repository = repository;
  }

  public async create(data: T): Promise<T> {
    const entity = await this.repository.create({ data });

    await prisma.$disconnect();

    return entity;
  }

  public async update<K>(id: K, data: T): Promise<T> {
    const entity = await this.repository.update({})

    await prisma.$disconnect();

    return entity;
  }

  public async delete<K>(id: K): Promise<void> {
    const entity = await this.repository.delete({})

    await prisma.$disconnect();
  }

  public async findOne<Q>(args: Q): Promise<T> {
    const entity = await this.repository.findUnique(args);

    await prisma.$disconnect();

    return entity;
  }

  public async findMany<Q>(args: Q): Promise<Pagination<T>> {
    const res = await Promise.all([
      this.repository.findMany(args),
      this.repository.count(args)
    ]);

    await prisma.$disconnect();

    return { content: res[0], count: res[1] };
  }
}

Thanks a lot, see ya.

**Comment by SevInf** [comment]
I think [type utilites](https://www.prisma.io/docs/orm/prisma-client/client-extensions/type-utilities#type-utilities) might be something you are looking for, in particaular `Prisma.Args` and `Prisma.Result`.  Let me know if that helps.

**Comment by SmashingQuasar** [comment]
> I think [type utilites](https://www.prisma.io/docs/orm/prisma-client/client-extensions/type-utilities#type-utilities) might be something you are looking for, in particaular `Prisma.Args` and `Prisma.Result`. Let me know if that helps.

I'm facing a similar issue as @MatheusFerreira71 and unfortunately it cannot be resolved with the current way Prisma handles types.
We are working on an application that has an `AbstractDelegatedRepository` that has a property `delegate` which must be any `Prisma.DelegateX`.
Since there is no `Prisma.AbstractDelegate` that is inherited by all delegates, it is impossible to properly type this property without having to manually maintain an `AbstractDelegate` which is not viable.

The solution you suggested @SevInf cannot work, since in this documentation, `T` must be determined in advance.
It's also a best practice when working with OOP and SOLID to use inheritance, this would solve so many issues.

**Comment by MatheusFerreira71** [comment]
That's right! Unfortunately, the Type Utilities were not very helpful in this situation.  

I was looking for something more like this:  

```typescript
export class UserRepository {
	constructor(private prisma: PrismaClient) {
		this.prisma = prisma;
	}

	async createUser(args: Prisma.UserCreateArgs): Promise<void> {
		this.prisma.user.create(args);
	}

	async readUser(args: Prisma.UserFindUniqueArgs): Promise<any> {
		return this.prisma.user.findUnique(args);
	}

	async readUsers(args: Prisma.UserFindManyArgs): Promise<any[]> {
		return this.prisma.user.findMany(args);
	}

	async updateUser(args: Prisma.UserUpdateArgs): Promise<void> {
		this.prisma.user.update(args);
	}

	async deleteUser(args: Prisma.UserDeleteArgs): Promise<void> {
		this.prisma.user.delete(args);
	}
}
```

This is a basic `UserRepository` class that manages user-related database operations.  
All methods utilize the Prisma types generated by `prisma generate`.  

Now, consider this:  

```typescript
export class Repository<T> {
	constructor(private prisma: Prisma.Delegate<T>) {
		this.prisma = prisma;
	}

	async create(args: Prisma.CreateArgs<T>): Promise<void> {
		this.prisma.create(args);
	}

	async read(args: Prisma.FindUniqueArgs<T>): Promise<any> {
		return this.prisma.findUnique(args);
	}

	async readMany(args: Prisma.FindManyArgs<T>): Promise<any[]> {
		return this.prisma.findMany(args);
	}

	async update(args: Prisma.UpdateArgs<T>): Promise<void> {
		this.prisma.update(args);
	}

	async delete(args: Prisma.DeleteArgs<T>): Promise<void> {
		this.prisma.delete(args);
	}
}
```

This is a generic `Repository` class intended to be extended by entity-specific repositories.  

Unlike the first example, which directly uses the Prisma-generated types, this version introduces an imaginary generic type to define a base repository structure. The generic type `T` represents the entity type—`User`, in this case.  

Having something like this would significantly improve adherence to OOP principles and SOLID design patterns. The same idea could be applied to return types, though I didn’t include them in this example.

**Comment by SmashingQuasar** [comment]
> That's right! Unfortunately, the Type Utilities were not very helpful in this situation.
> 
> I was looking for something more like this:
> 
> export class UserRepository {
> 	constructor(private prisma: PrismaClient) {
> 		this.prisma = prisma;
> 	}
> 
> 	async createUser(args: Prisma.UserCreateArgs): Promise<void> {
> 		this.prisma.user.create(args);
> 	}
> 
> 	async readUser(args: Prisma.UserFindUniqueArgs): Promise<any> {
> 		return this.prisma.user.findUnique(args);
> 	}
> 
> 	async readUsers(args: Prisma.UserFindManyArgs): Promise<any[]> {
> 		return this.prisma.user.findMany(args);
> 	}
> 
> 	async updateUser(args: Prisma.UserUpdateArgs): Promise<void> {
> 		this.prisma.user.update(args);
> 	}
> 
> 	async deleteUser(args: Prisma.UserDeleteArgs): Promise<void> {
> 		this.prisma.user.delete(args);
> 	}
> }
> 
> This is a basic `UserRepository` class that manages user-related database operations. All methods utilize the Prisma types generated by `prisma generate`.
> 
> Now, consider this:
> 
> export class Repository<T> {
> 	constructor(private prisma: Prisma.Delegate<T>) {
> 		this.prisma = prisma;
> 	}
> 
> 	async create(args: Prisma.CreateArgs<T>): Promise<void> {
> 		this.prisma.create(args);
> 	}
> 
> 	async read(args: Prisma.FindUniqueArgs<T>): Promise<any> {
> 		return this.prisma.findUnique(args);
> 	}
> 
> 	async readMany(args: Prisma.FindManyArgs<T>): Promise<any[]> {
> 		return this.prisma.findMany(args);
> 	}
> 
> 	async update(args: Prisma.UpdateArgs<T>): Promise<void> {
> 		this.prisma.update(args);
> 	}
> 
> 	async delete(args: Prisma.DeleteArgs<T>): Promise<void> {
> 		this.prisma.delete(args);
> 	}
> }
> 
> This is a generic `Repository` class intended to be extended by entity-specific repositories.
> 
> Unlike the first example, which directly uses the Prisma-generated types, this version introduces an imaginary generic type to define a base repository structure. The generic type `T` represents the entity type—`User`, in this case.
> 
> Having something like this would significantly improve adherence to OOP principles and SOLID design patterns. The same idea could be applied to return types, though I didn’t include them in this example.
> 
> And we could still leverage the types generated by `prisma generate` by using **indexed type mapping**.
> 
> For example:
> 
> type ModelFindUniqueArgsMap = {
>   User: UserFindUniqueArgs;
> };
> 
> type FindUniqueArgs<T extends keyof ModelFindUniqueArgsMap> = ModelFindUniqueArgsMap[T];
> 
> type MyArgs = FindUniqueArgs<"User">;
> 
> Since Prisma types follow a consistent naming convention, we could create a mapping type where each key represents a database entity as a string, and the corresponding value is the Prisma-generated type for that entity.

I vehemently agree with everything that you said.

The current approach of Prisma is very intensive and repetitive in types. Using inheritance would have the side effect benefit of drastically reducing the size of the exported types.
